File: src/languages/cpp.rs
pub mod cpp {
    use crate::language::language::{LanguageInformation, Language, execute_command};
    use crate::program_output::program_output::ProgramOutput;
    use crate::compiled_language::compiled_language::CompiledLanguage;
    use crate::config::config::get_work_dir;

    pub struct Cpp {
        info: LanguageInformation,
        executable_path: String,
    }

    impl Language for Cpp {
        fn from_config() -> Result<Self, String> where Self: Sized {
            let language_info_from_config: Result<LanguageInformation, String> = LanguageInformation::from_config("cpp");

            let executable_path: String = get_work_dir().unwrap() + "/a.out";
            return match language_info_from_config {
                Ok(info) => Ok(
                    Cpp {
                        info,
                        executable_path,
                    }),
                Err(err) => Err(err)
            };
        }

        fn get_language_information(&self) -> &LanguageInformation {
            &self.info
        }

        fn run(&self, code_path: &str) -> Result<ProgramOutput, String> {
            match self.compile(code_path) {
                Ok(_) => execute_command(self.get_executable_path(), vec!()),
                Err(err) => Err(err)
            }
        }
    }

    impl CompiledLanguage for Cpp {
        fn get_executable_path(&self) -> &str {
            self.executable_path.as_str()
        }

        fn compile(&self, code_path: &str) -> Result<ProgramOutput, String> {
            execute_command(self.get_language_information().get_executable(),
                            vec![code_path,
                                 "-o",
                                 self.get_executable_path()])
        }
    }
}

File: src/languages/cpp.py
"""
Module containing the C++ language class.
"""
import os
from subprocess import CompletedProcess

from src.compiled_language import CompiledLanguage
from src.config import Config
from src.language import execute_command
from src.language_information import LanguageInformation


class Cpp(CompiledLanguage):
    """
    Class representing the C++ language.
    """
    def __init__(self, config: Config):

        language_info = config.create_language_information('cpp')

        self.__info: LanguageInformation = language_info
        self.__compiled_output = config.get_item('languages')['cpp']['run']

    def get_language_information(self) -> LanguageInformation:
        return self.__info

    def run(self, code_path: str) -> CompletedProcess:
        self.compile(code_path)

        return execute_command(self.get_executalbe_path(), [])

    def compile(self, code_path: str) -> CompletedProcess:
        """
        Compiles the code at the given path

        Returns:
            CompletedProcess: The result of the compilation
        """
        return execute_command(self.__info.executable,
                               [code_path, '-o', self.get_executalbe_path()])

    def get_executalbe_path(self) -> str:
        """
        Returns the path to the executable
        """
        return os.path.join(Config().get_item('work_dir'), self.__compiled_output)
